app/celery/process_letter_client_response_tasks.py

```diff
@@ -0,0 +1,56 @@
+from flask import current_app
+
+from app import notify_celery
+from app.constants import (
+    DVLA_TO_NOTIFICATION_STATUS_MAP,
+    NOTIFICATION_TECHNICAL_FAILURE,
+)
+from app.dao.notifications_dao import (
+    dao_get_notification_or_history_by_id,
+    dao_update_notification,
+)
+from app.exceptions import NotificationTechnicalFailureException
+
+
+@notify_celery.task(bind=True, name="process-letter-callback")
+def process_letter_callback_data(self, notification_id, page_count, dvla_status):
+    notification = dao_get_notification_or_history_by_id(notification_id)
+
+    validate_billable_units(notification, page_count)
+
+    new_status = determine_new_status(dvla_status)
+
+    if is_duplicate_update(notification.status, new_status):
+        current_app.logger.info(
+            "Duplicate update received for notification id: %s with status: %s",
+            notification_id,
+            new_status,
+        )
+        return
+
+    notification.status = new_status
+
+    dao_update_notification(notification)
+
+    if new_status == NOTIFICATION_TECHNICAL_FAILURE:
+        raise NotificationTechnicalFailureException(
+            f"Letter status received as REJECTED for notification id: {notification.id}"
+        )
+
+
+def validate_billable_units(notification, dvla_page_count):
+    if notification.billable_units != int(dvla_page_count):
+        current_app.logger.error(
+            "Notification with id %s has %s billable_units but DVLA says page count is %s",
+            notification.id,
+            notification.billable_units,
+            dvla_page_count,
+        )
+
+
+def determine_new_status(dvla_status):
+    return DVLA_TO_NOTIFICATION_STATUS_MAP[dvla_status]
+
+
+def is_duplicate_update(current_status, new_status):
```

Comment:
Do we want to also check for the case where we receive a second, different, callback status for a notification? It _shouldn't_ ever happen, but that would be something that we'd want to know about. E.g. when updating other notifications, we [do it this way](https://github.com/alphagov/notifications-api/blob/6f79e083cfe3cf6075c92013a23b8bd027a80dd5/app/dao/notifications_dao.py#L177-L185)

Reply:
https://github.com/alphagov/notifications-api/pull/4157/commits/57d86ab061e8012cca513b6f6cd4de8afe56ea0e

I wanted to reuse the method `_duplicate_update_warning` - however, I noticed that if we have the underscore at the beginning it means it is a private method.

I think it would be nice to use this method for consistency in logs, do you recommend me renaming the method to remove the underscore or to duplicate the function?

Reply:
ok I will address the refactoring at a separate PR and just leave the current implementation as is